### PR TITLE
Fix reading EstEID certificates with T=0

### LINKED ISSUE #1190: esteid: OpenSC fails to read file with T=0 recursive APDU-s
### Problem Description
OpenSC fails to read file with T=0

### Proposed Resolution
Instead sending `00 B0 01 00 00` it should send `00 C0 00 00 00` because last APDU responded `61 00`

### Steps to reproduce
force protocol to T=0 in config

### Logs
```
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] card-mcrd.c:811:do_select: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] card-mcrd.c:861:select_part: select_part (0xAACE, kind=2)
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] apdu.c:554:sc_transmit_apdu: called
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] card.c:407:sc_lock: called
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] apdu.c:521:sc_transmit: called
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:0, INS:A4, P1:2, P2:4, data(2) 0x7ffeefbfc4b6
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] reader-pcsc.c:283:pcsc_transmit: reader 'ACS ACR 38U-CCID'
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (7 bytes):
00 A4 02 04 02 AA CE .......
0x7fffa9a9e340 13:18:08.195 [opensc-pkcs11] reader-pcsc.c:212:pcsc_internal_transmit: called
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (2 bytes):
61 10 a.
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] apdu.c:434:sc_get_response: called
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] apdu.c:554:sc_transmit_apdu: called
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] card.c:407:sc_lock: called
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] apdu.c:521:sc_transmit: called
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:0, INS:C0, P1:0, P2:0, data(0) 0x0
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] reader-pcsc.c:283:pcsc_transmit: reader 'ACS ACR 38U-CCID'
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 C0 00 00 10 .....
0x7fffa9a9e340 13:18:08.247 [opensc-pkcs11] reader-pcsc.c:212:pcsc_internal_transmit: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (18 bytes):
62 0E 82 01 01 83 02 AA CE 85 02 08 00 8A 01 05 b...............
90 00                                           ..
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:505:sc_get_response: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card-mcrd.c:655:process_fcp: processing FCI bytes
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card-mcrd.c:662:process_fcp:   file identifier: 0xAACE
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card-mcrd.c:677:process_fcp:   bytes in file: 2048
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card-mcrd.c:701:process_fcp:   shareable: no
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card-mcrd.c:721:process_fcp:   type: working EF
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card-mcrd.c:723:process_fcp:   EF structure: 1
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card-mcrd.c:1154:mcrd_select_file:   result=0, ef=1, curpath=3F00EEEEAACE
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:783:sc_select_file: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:565:sc_read_binary: called; 2048 bytes at index 0
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:407:sc_lock: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:565:sc_read_binary: called; 256 bytes at index 0
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:554:sc_transmit_apdu: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:407:sc_lock: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:521:sc_transmit: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:0, INS:B0, P1:0, P2:0, data(0) 0x0
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] reader-pcsc.c:283:pcsc_transmit: reader 'ACS ACR 38U-CCID'
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 B0 00 00 00 .....
0x7fffa9a9e340 13:18:08.253 [opensc-pkcs11] reader-pcsc.c:212:pcsc_internal_transmit: called
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (258 bytes):
30 82 06 47 30 82 04 2F A0 03 02 01 02 02 10 73 0..G0../.......s
3F 10 6C 71 63 F0 89 59 D4 C5 C0 32 80 D6 34 30 ?.lqc..Y...2..40
0D 06 09 2A 86 48 86 F7 0D 01 01 0B 05 00 30 6B ...*.H........0k
31 0B 30 09 06 03 55 04 06 13 02 45 45 31 22 30 1.0...U....EE1"0
20 06 03 55 04 0A 0C 19 41 53 20 53 65 72 74 69  ..U....AS Serti
66 69 74 73 65 65 72 69 6D 69 73 6B 65 73 6B 75 fitseerimiskesku
73 31 17 30 15 06 03 55 04 61 0C 0E 4E 54 52 45 s1.0...U.a..NTRE
45 2D 31 30 37 34 37 30 31 33 31 1F 30 1D 06 03 E-107470131.0...
55 04 03 0C 16 54 45 53 54 20 6F 66 20 45 53 54 U....TEST of EST
45 49 44 2D 53 4B 20 32 30 31 35 30 1E 17 0D 31 EID-SK 20150...1
37 31 30 30 34 31 31 32 38 30 30 5A 17 0D 32 32 71004112800Z..22
30 39 32 30 32 30 35 39 35 39 5A 30 81 9B 31 0B 0920205959Z0..1.
30 09 06 03 55 04 06 13 02 45 45 31 0F 30 0D 06 0...U....EE1.0..
03 55 04 0A 0C 06 45 53 54 45 49 44 31 17 30 15 .U....ESTEID1.0.
06 03 55 04 0B 0C 0E 61 75 74 68 65 6E 74 69 63 ..U....authentic
61 74 69 6F 6E 31 26 30 24 06 03 55 04 03 0C 1D ation1&0$..U....
61 00                                           a.
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] apdu.c:434:sc_get_response: called
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] apdu.c:554:sc_transmit_apdu: called
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] card.c:407:sc_lock: called
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] apdu.c:521:sc_transmit: called
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:0, INS:C0, P1:0, P2:0, data(0) 0x0
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] reader-pcsc.c:283:pcsc_transmit: reader 'ACS ACR 38U-CCID'
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 C0 00 00 00 .....
0x7fffa9a9e340 13:18:08.316 [opensc-pkcs11] reader-pcsc.c:212:pcsc_internal_transmit: called
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (258 bytes):
4D C3 84 4E 4E 49 4B 2C 4D 41 52 49 2D 4C 49 49 M..NNIK,MARI-LII
53 2C 36 31 37 30 39 32 31 30 30 30 35 31 10 30 S,617092100051.0
0E 06 03 55 04 04 0C 07 4D C3 84 4E 4E 49 4B 31 ...U....M..NNIK1
12 30 10 06 03 55 04 2A 0C 09 4D 41 52 49 2D 4C .0...U.*..MARI-L
49 49 53 31 14 30 12 06 03 55 04 05 13 0B 36 31 IIS1.0...U....61
37 30 39 32 31 30 30 30 35 30 76 30 10 06 07 2A 7092100050v0...*
86 48 CE 3D 02 01 06 05 2B 81 04 00 22 03 62 00 .H.=....+...".b.
04 46 A6 62 03 34 EF 87 E6 A5 E0 AC 11 5D 63 27 .F.b.4.......]c'
86 4F F6 5F EB B1 50 1F 0E 6F B1 3C 07 E5 75 91 .O._..P..o.<..u.
FC 3B AF 51 C9 14 38 FE 4A FD 56 8B 7E 67 04 DE .;.Q..8.J.V.~g..
64 39 A9 0B 8B E5 16 9D 4E 43 A8 71 EC 89 FA 3A d9......NC.q...:
D0 F3 70 CC E7 45 15 5A 45 9C 80 67 B8 DF F8 D9 ..p..E.ZE..g....
E7 E4 56 D1 86 42 A0 34 88 86 C1 D5 AC 85 09 1D ..V..B.4........
04 A3 82 02 62 30 82 02 5E 30 09 06 03 55 1D 13 ....b0..^0...U..
04 02 30 00 30 0E 06 03 55 1D 0F 01 01 FF 04 04 ..0.0...U.......
03 02 03 88 30 81 89 06 03 55 1D 20 04 81 81 30 ....0....U. ...0
61 00                                           a.
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:505:sc_get_response: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] iso7816.c:171:iso7816_read_binary: returning with: 256
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] card.c:605:sc_read_binary: returning with: 256
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] card.c:565:sc_read_binary: called; 256 bytes at index 256
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:554:sc_transmit_apdu: called
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] card.c:407:sc_lock: called
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:521:sc_transmit: called
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:0, INS:B0, P1:1, P2:0, data(0) 0x0
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] reader-pcsc.c:283:pcsc_transmit: reader 'ACS ACR 38U-CCID'
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 B0 01 00 00 .....
0x7fffa9a9e340 13:18:08.345 [opensc-pkcs11] reader-pcsc.c:212:pcsc_internal_transmit: called
0x7fffa9a9e340 13:18:08.467 [opensc-pkcs11] reader-pcsc.c:293:pcsc_transmit: 
Incoming APDU (2 bytes):
6F 00 o.
```

### PR DESCRIPTION
**Pull request by metsma**
Fixes #1190

Signed-off-by: Raul Metsma <raul@metsma.ee>

##### Checklist

- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [X] Tested with the following card: EstEID
	- [X] tested PKCS#11
	- [ ] tested Windows Minidriver
	- [ ] tested macOS Tokend

### CODE REVIEW

**Comment by frankmorgner** [comment]
Why do you need the global flag `SC_CARD_CAP_READ_BINARY_NO_BREAK` if you solve the problem with GET RESPONSE at the card driver level?

**Comment by metsma** [comment]
https://github.com/OpenSC/OpenSC/issues/1190#issuecomment-343150779
As dengert suggested to avoid spliting sc_read_binary

> if (count > max_le && !(card->caps & SC_CARD_CAP_READ_BINARY_NO_BREAK)) {
Then your version of read_binary would get control with the original count and would simplify your drivers version of read_binary.
> So it looks like you need a read_binary routine in your card driver. The problem is sc_read_binary will try and break up any read_binary into multiple read_binary commands where the second one will have an index != 0. sc_read_binary before you get control. max_le = sc_get_max_recv_size(card);

**Comment by dengert** [comment]
@frankmorgner  The name SC_CARD_CAP_READ_BINARY_NO_BREAK could be changed if you want. 
I was surprised how many card drivers have their own version of read_binary.  read_binary  is not a standard as one might expect.  The SC_CARD_CAP_READ_BINARY_NO_BREAK tells sc_read_binary to just call the card's version.

**Comment by metsma** [comment]
Any suggestion? I copied from dengert comment

**Comment by frankmorgner** [comment]
A long time ago I thought about putting all this GET RESPONSE magic from apdu.c into iso7816.c. If i remember correctly, there are many more extensions from ISO 7816 that seeped through to the generic layers of OpenSC. With such separation, you would have an easier job of modifying the card specific behavior before some ISO 7816 mechanism kicks in. I think this would be the *right way* to go in terms of establishing long term software quality. However, this has a lot of potential to break things. 

Introducing more complexity (read: hacks) doesn't solve the problem in the long run. I'd suggest to either start working on the refactoring now, which could then be included in 0.19.0 or to solve this problem exclusively on the card driver level.

**Comment by dengert** [comment]
The problem the card driver does not get control until after sc_read_binary has screwed thing up. 

And sc_read_binary is called from  non-card driver files card-*.c, pkcs15-*.c and pkcs15init/* ./libopensc/pkcs15.c calls it 5 times, ./libopensc/dir.c, /libopensc/iso7816.c, ./libopensc/card.c, ./tools/opensc-tool.c, ./tools/netkey-tool.c, 
./tools/cryptoflex-tool.c, ./tools/westcos-tool.c and ./tools/opensc-explorer.c.    
It is also called from card driver files: card-*.c, pkcs15-*.c and pkcs15init/*

So, refracting looks like a big project.

**Comment by frankmorgner** [comment]
What happens if you use `max_recv_len = 255` in `opensc.conf` (or in the code for the card's `read_binary`)? Does the card still respond with `61 00`?

**Comment by dengert** [comment]
Is this an issue with T=0 and T=1 or the reader. IIRC, with T=0 the card does not see the Le, only the 
reader does. Would a PCSC trace show more?  

https://github.com/metsma, to you have another reader that does T=0 , or can you force T=0 to see if it fails or works?

**Comment by metsma** [comment]
Yes seems like it helps
Default: https://www.dropbox.com/s/4oavydmuuxm4hc2/opensc-debug.log.broken?dl=0
Modified config: https://www.dropbox.com/s/igfqkp4pdp9f5rl/opensc-debug.log.config?dl=0

File: src/libopensc/card-mcrd.c [hunk 1]
```diff
@@ -347,15 +347,18 @@ static int mcrd_init(sc_card_t * card)
...
-	                        apdu.lc = sizeof(EstEID_v35_AID);
-        	                apdu.data = EstEID_v35_AID;
-                	        apdu.datalen = sizeof(EstEID_v35_AID);
-                        	apdu.resplen = 0;
-	                        apdu.le = 0;
+							apdu.lc = sizeof(EstEID_v35_AID);
+							apdu.data = EstEID_v35_AID;
+							apdu.datalen = sizeof(EstEID_v35_AID);
+							apdu.resplen = 0;
+							apdu.le = 0;
```
Comment (frankmorgner) [inline src/libopensc/card-mcrd.c]: please fix the indenting

File: src/libopensc/card-mcrd.c [hunk 2]
```diff
@@ -347,15 +347,18 @@ static int mcrd_init(sc_card_t * card)
...
 				r = sc_transmit_apdu(card, &apdu);
-	                        SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "APDU transmit failed");
-        	                sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "SELECT AID: %02X%02X", apdu.sw1, apdu.sw2);
-				if (apdu.sw1 != 0x90 && apdu.sw2 != 0x00) {
+							SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "APDU transmit failed");
+							sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "SELECT AID: %02X%02X", apdu.sw1, apdu.sw2);
+				if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00) {
+					if (card->reader && card->reader->active_protocol == SC_PROTO_T0)
+						card->max_recv_size = 255;
```
Comment (frankmorgner) [inline src/libopensc/card-mcrd.c]: please add a comment why you're doing the workaround

File: src/libopensc/card-mcrd.c [hunk 3]
```diff
@@ -347,15 +347,18 @@ static int mcrd_init(sc_card_t * card)
...
+							apdu.lc = sizeof(EstEID_v35_AID);
+							apdu.data = EstEID_v35_AID;
+							apdu.datalen = sizeof(EstEID_v35_AID);
+							apdu.resplen = 0;
+							apdu.le = 0;
 				r = sc_transmit_apdu(card, &apdu);
-	                        SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "APDU transmit failed");
-        	                sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "SELECT AID: %02X%02X", apdu.sw1, apdu.sw2);
-				if (apdu.sw1 != 0x90 && apdu.sw2 != 0x00) {
+							SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "APDU transmit failed");
```
Comment (frankmorgner) [inline src/libopensc/card-mcrd.c]: indenting

File: src/libopensc/card-mcrd.c (hunk 1 above)
Comment (metsma) [reply src/libopensc/card-mcrd.c]: It is fixing indenting, before there was spaces and tabs mixed

Comment (metsma) [reply src/libopensc/card-mcrd.c]: or you mean leave the mixed indenting as it is?

File: src/libopensc/card-mcrd.c (hunk 2 above)
Comment (metsma) [reply src/libopensc/card-mcrd.c]: Same here, spaces and tabs mixed before and do I leave as it is?

File: src/libopensc/card-mcrd.c (hunk 3 above)
Comment (metsma) [reply src/libopensc/card-mcrd.c]: Same here, spaces and tabs mixed before and do I leave as it is?

File: src/libopensc/card-mcrd.c (hunk 1 above)
Comment (frankmorgner) [reply src/libopensc/card-mcrd.c]: There are too many tabs. Look at GitHub's diff, you'll see the problem.

**Comment by frankmorgner** [comment]
thanks

**Pull request merged**